influxql/engine.go

```diff
@@ -131,14 +123,30 @@ func (m *MapReduceJob) Execute(out chan *Row, filterEmptyResults bool) {
 
 			return
 		} else {
-			limit := m.stmt.Limit
-			if m.stmt.Offset+m.stmt.Limit > len(resultValues) {
-				limit = len(resultValues) - m.stmt.Offset
+			pointCountInResult = m.stmt.Limit
+			if m.stmt.Offset+m.stmt.Limit > pointCountInResult {
+				pointCountInResult = pointCountInResult - m.stmt.Offset
 			}
-
-			resultTimes = resultTimes[m.stmt.Offset : m.stmt.Offset+limit]
-			resultValues = resultValues[m.stmt.Offset : m.stmt.Offset+limit]
 		}
+	}
+
+	// initialize the times of the aggregate points
+	resultTimes := make([]int64, pointCountInResult)
+	resultValues := make([][]interface{}, pointCountInResult)
+
+	// ensure that the start time for the results is on the start of the window
+	startTimeBucket := m.TMin / m.interval * m.interval
+
+	for i, _ := range resultTimes {
+		t := startTimeBucket + (int64(i) * m.interval * int64(m.stmt.Offset+1))
+		resultTimes[i] = t
+		// we always include time so we need one more column than we have aggregates
+		vals := make([]interface{}, 0, len(aggregates)+1)
+		resultValues[i] = append(vals, time.Unix(0, t).UTC())
+	}
+
+	// set the minimum time if we should set the limit and the offset is greater than 0. This way we don't range over data we don't need to
+	if setLimit && m.stmt.Offset > 0 {
```

Comment:
OK. The comment is a little unclear since "set the limit" sounds like the code should set some limit, as opposed to the fact the client had a limit in the query. I'm not sure if even I interpreted the comment correctly.